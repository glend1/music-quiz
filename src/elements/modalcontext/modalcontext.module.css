#modal {
	display: none;
	backdrop-filter: blur(1rem) brightness(80%);
	min-height: 100%;
	width: 100vw;
	position: absolute;
	top: 0px;
	left: 0px;
}

#modal.unhide {
	display: flex;
	justify-content: center;
	align-items: center;
	flex-wrap: wrap;
}

.card {
	background-color: var(--bg-colour);
	border-radius: 0.3rem;
	border: 1px solid var(--outline-colour);
	padding: 1rem;
	margin: 1rem;
	width: 30%;
	min-height: 50%;
}

.card button {
	float: right;
}
