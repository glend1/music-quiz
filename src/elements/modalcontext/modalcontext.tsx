import { createContext, MouseEventHandler, useContext } from "react";
import * as ReactDOM from "react-dom/client";
import styles from "./modalcontext.module.css";

type modalTypes = {
	scale?: string;
	chord?: string;
};

function Modal() {
	function openModal(cb: (data: modalTypes) => JSX.Element) {
		const returnFunction: MouseEventHandler<HTMLAnchorElement | HTMLElement> = (
			e
		) => {
			let modalElement = document.getElementById(styles.modal);
			let data: modalTypes = (e.target as HTMLElement).dataset;
			if (modalElement) {
				let div = document.createElement("div");
				div.classList.add(styles.card);
				div.addEventListener("click", (e) => {
					e.stopPropagation();
				});
				modalElement.appendChild(div);
				ReactDOM.createRoot(div).render(
					<>
						<div>{cb(data)}</div>
						<div className={styles.close}>
							<button onClick={closeModal}>Close</button>
						</div>
					</>
				);
				if (!modalElement.classList.contains(styles.unhide)) {
					modalElement.classList.add(styles.unhide);
				}
			}
		};
		return returnFunction;
	}
	let closeModal: MouseEventHandler<HTMLElement> = (e) => {
		let modalElement = document.getElementById(styles.modal);
		if (modalElement) {
			let elem = (e.target as HTMLElement).parentNode?.parentNode;
			if (elem) {
				modalElement.removeChild(elem);
			}
			if (modalElement.childElementCount < 1) {
				modalElement.classList.remove(styles.unhide);
			}
		}
	};
	let closeModalGlobal: MouseEventHandler<HTMLElement> = (e) => {
		let modalElement = document.getElementById(styles.modal);
		if (modalElement) {
			modalElement.innerHTML = "";
			modalElement.classList.remove(styles.unhide);
		}
	};
	return {
		openModal,
		closeModalGlobal,
	};
}

const modalContext = createContext(Modal().openModal);

export function GlobalModalContext({
	children,
	modal = false,
}: {
	children?: React.ReactNode;
	modal?: boolean;
}) {
	const modalElement = Modal();
	return (
		<modalContext.Provider value={modalElement.openModal}>
			{children}
			{modal ? (
				<div id={styles.modal} onClick={modalElement.closeModalGlobal}></div>
			) : null}
		</modalContext.Provider>
	);
}

export function useModalContext() {
	return useContext(modalContext);
}
