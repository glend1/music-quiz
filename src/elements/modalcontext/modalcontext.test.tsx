import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { GlobalModalContext, useModalContext } from "./modalcontext";

function ModalMock() {
	return (
		<GlobalModalContext modal={true}>
			<OpenModalMock />
		</GlobalModalContext>
	);
}

function OpenModalMock() {
	const context = useModalContext();
	return (
		<button
			onClick={context((data) => {
				return <h3>Rendered</h3>;
			})}
		>
			Open Modal
		</button>
	);
}

describe("GlobalModalContext", () => {
	it("Should render the component", () => {
		const html = render(
			<GlobalModalContext modal={true}>
				<h1>Hello World</h1>
			</GlobalModalContext>
		);
		expect(html.container).toMatchSnapshot();
	});
	it("Should not render the modal", () => {
		const html = render(
			<GlobalModalContext>
				<h1>Hello World</h1>
			</GlobalModalContext>
		);
		expect(html.container).toMatchSnapshot();
	});
});
describe("useModalContext", () => {
	it("Should open the modal", async () => {
		const html = render(<ModalMock />);
		await userEvent.click(screen.getByRole("button"));
		expect(html.container).toMatchSnapshot();
	});
	it("Should close the modal", async () => {
		const html = render(<ModalMock />);
		await userEvent.click(screen.getByRole("button"));
		await userEvent.click(screen.getAllByRole("button")[1]);
		expect(html.container).toMatchSnapshot();
	});
	it("Should close the modal globally", async () => {
		const html = render(<ModalMock />);
		await userEvent.click(screen.getByRole("button"));
		const modal = html.container.querySelector("#modal");
		if (modal) {
			await userEvent.click(modal);
		}
		expect(html.container).toMatchSnapshot();
	});
});
