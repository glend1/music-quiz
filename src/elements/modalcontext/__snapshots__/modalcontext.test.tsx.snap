// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalModalContext Should not render the modal 1`] = `
<div>
  <h1>
    Hello World
  </h1>
</div>
`;

exports[`GlobalModalContext Should render the component 1`] = `
<div>
  <h1>
    Hello World
  </h1>
  <div
    id="modal"
  />
</div>
`;

exports[`useModalContext Should close the modal 1`] = `
<div>
  <button>
    Open Modal
  </button>
  <div
    class=""
    id="modal"
  />
</div>
`;

exports[`useModalContext Should close the modal globally 1`] = `
<div>
  <button>
    Open Modal
  </button>
  <div
    class=""
    id="modal"
  />
</div>
`;

exports[`useModalContext Should open the modal 1`] = `
<div>
  <button>
    Open Modal
  </button>
  <div
    class="unhide"
    id="modal"
  >
    <div
      class="card"
    >
      <h3>
        Rendered
      </h3>
      <button>
        Close
      </button>
    </div>
  </div>
</div>
`;
